2-advanced-topics/data-pipelines/intro_data_pipelines.ipynb

```diff
@@ -0,0 +1,1037 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "233863f5-14ef-448d-b89e-d532c64a160c",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "outputs": [],
+   "source": [
+    "pop_df = spark.read.table(f\"{database_name}.kenya_subnational_population\").alias(\"pop_df\")\n",
+    "\n",
+    "agg_df = (spark.read.table(f\"{database_name}.kenya_gold\")\n",
+    "    .groupBy(\"country_name\", \"year\", \"geo1\", \"func\",)\n",
+    "    .agg(sum(\"executed\").alias(\"expenditure\"))\n",
+    ").alias(\"agg_df\")\n",
+    "\n",
+    "agg_with_pop_df = (agg_df\n",
+    "    .join(pop_df, \n",
+    "          (agg_df.country_name == pop_df.country_name) & \n",
+    "          (agg_df.geo1 == pop_df.adm1_name) & \n",
+    "          (agg_df.year == pop_df.year), \n",
+    "          \"inner\")\n",
+    "    .select(\"agg_df.country_name\", \"agg_df.year\", \"agg_df.geo1\", \"agg_df.func\", \"agg_df.expenditure\", \"pop_df.population\")\n",
+    ")\n",
+    "\n",
+    "agg_with_pop_df = agg_with_pop_df.withColumn(\n",
+    "    'per_capita_spending',\n",
+    "    (col('expenditure') / col('population')).alias('per_capita_spending')\n",
+    ")\n",
+    "\n",
+    "agg_with_pop_df.write.format(\"delta\").mode(\"overwrite\").saveAsTable(f\"{database_name}.kenya_func_geo1_agg\")\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "d122eee1-a4a3-4526-9d77-fdd05afa0238",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "# Orchestration\n",
+    "\n",
+    "Orchestration refers to the process of coordinating and automating the various stages of the pipeline—such as data extraction, transformation, and loading—to ensure smooth, sequential execution. This is equivalent to *connecting the individual jobs* so that they run in an organized manner keeping track of the dependencies between the jobs. In PySpark, this involves managing the flow of tasks.\n",
+    "\n",
+    "Orchestration tools help **schedule** and manage these stages, ensuring that dependencies between tasks are handled correctly and that the pipeline runs efficiently. This ensures data moves through the pipeline in an automated manner, handling errors and retries where necessary.\n",
+    "\n",
+    "We will use the databricks built-in *Workflows* for managing these dependencies and sceduling. "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "application/vnd.databricks.v1+cell": {
+     "cellMetadata": {},
+     "inputWidgets": {},
+     "nuid": "a3ef6f62-8d24-4c15-a221-95d5be0a1dba",
+     "showTitle": false,
+     "title": ""
+    }
+   },
+   "source": [
+    "![Medallion Flowchart](assets/kenya_workflow.png)\n"
```

Comment:
note that this image is not in the repository. You can upload it or send it to me before the session and I'll do it